src/cephadm/cephadm

```diff
@@ -3450,26 +3453,144 @@ def command_list_networks():
-    host_version = None
-    ls = []
+def get_enabled_ceph_units(fsid: str) -> List[str]:
+    try:
+        units = os.listdir(f'/etc/systemd/system/ceph-{fsid}.target.wants')
+    except FileNotFoundError:
+        return []
+    else:
+        return units
+
 
-    data_dir = args.data_dir
+def get_unit_states(fsid: str) -> Dict[str, Dict[str, str]]:
+    cmd = f'systemctl --all --no-legend -t service list-units ceph-{fsid}*'
+    unit_states = {}
+    std_out, std_err, rc = call(cmd.split(' '), verbose_on_failure=False)
+    if not rc:
+        for line in std_out.split('\n'):
+            if not line:
+                continue
+            unit_name, load, active, sub = line.split()[0:4]
+            unit_states[unit_name] = {
+                "load": load,
+                "active": active,
+                "sub": sub,
+            }
+    return unit_states
+
+
+def list_daemons(detail=True, legacy_dir=None, target_fsid=None):
+    # type: (bool, Optional[str], Optional[str]) -> List[Dict[str, str]]
+    host_version: Optional[str] = None
+    ls: List[Dict[str, Optional[Union[str, bool]]]] = []
+    fsid: str
+    info: Dict[str, Optional[Union[str, bool]]]
+
+
+    def _http_call(url: str, context: Optional[ssl.SSLContext] = None) -> str:
+        try:
+            with urlopen(url, context=context) as response:
+                return response.read().decode('utf-8')
+        except URLError:
+            return ''
+
+    def _get_listen_address(file_name: str) -> Optional[str]:
+        with open(file_name) as f:
+            data = f.read().strip()
+        if "web.listen-address" in data:
+            listen_address = data.split('--web.listen-address=')[1].split(' ')[0]
+            return listen_address.split(':')[-1]
+        return None
+
+    def _get_grafana_port(file_name: str) -> Optional[str]:
+        with open(file_name) as f:
+            data = f.read().strip()
+        for line in data.split('\n'):
+            if 'http_port' in line and line.startswith(' '):
+                return line.split('=')[1].strip()
+        return None
+
+    def _get_web_port(service: str) -> Optional[str]:
+        dirs = os.listdir(os.path.join(data_dir, fsid))
+        for dir_name in dirs:
+            if dir_name.startswith(service):
+                if service in ['node-exporter', 'alertmanager', 'prometheus']:
+                    return _get_listen_address(os.path.join(data_dir, fsid, dir_name, 'unit.run'))
+                elif service == 'grafana':
+                    return _get_grafana_port(os.path.join(data_dir, fsid, dir_name, 'etc', 'grafana', 'grafana.ini'))
+                break
+        return None
+
+    def _get_grafana_version(default_port: str = '3000') -> str:
+        port = _get_web_port('grafana') or default_port
+        ctx = ssl.create_default_context()
+        ctx.check_hostname = False
+        ctx.verify_mode = ssl.CERT_NONE
+        data = _http_call(
```

Comment:
When grafana/prometheus/node exporter  daemons starts there is a time where the health/status/metrics api endpoint does not return any information (data=None). If you execute the ls command in this moment your are going to get an error like this: (this is in grafana, but it also happens in prometheus and node exporter)

```
[root@cephLab2-node-00 bin]# cephadm ls
Traceback (most recent call last):
  File "/root/bin/cephadm", line 5929, in <module>
    r = args.func()
  File "/root/bin/cephadm", line 3458, in command_ls
    ls = list_daemons(detail=not args.no_detail,
  File "/root/bin/cephadm", line 3741, in list_daemons
    version = method()  # type: ignore
  File "/root/bin/cephadm", line 3555, in _get_grafana_version
    return json.loads(data)['version']
  File "/usr/lib64/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/lib64/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib64/python3.8/json/decoder.py", line 355, in raw_decode
    raise JSONDecodeError("Expecting value", s, err.value) from None
json.decoder.JSONDecodeError: Expecting value: line 1 column 1 (char 0)
```

So it seems that the three methods to get versions from the monitoring stack needs a check before trying to do something with data.

Reply:
ACK